### PR TITLE
Allocate continuous row id for  single INSERT statement

### PR DESCRIPTION
**Pull request by Little-Wallace**
### What problem does this PR solve? 

If a table is created without primary key, TiDB will assign a unique RowID to every record and store it in TiKV. This PR will make the RowID of every statement be continuous. 
The continuous keys are good for TiKV to improve write performance,  by skip checking write version, see  [TiKV #5846](https://github.com/tikv/tikv/pull/5846/files) for more details.

 For example, we run this sql in TiDB:
```
CREATE TABLE t1 (
  `k1` bigint(64) DEFAULT NULL,
  `k2` bigint(64) DEFAULT NULL,
  `v1` varchar(300) DEFAULT NULL,
  `v2` int(11) NOT NULL,
   KEY `k1`(`k1`)
);
thread1: INSERT INTO t1(k1,k2,v1,v2) VALUES(1,2,'a','a'),(2,2,'a','a');
thread2: INSERT INTO t1(k1,k2,v1,v2) VALUES(2,2,'a','a'),(3,2,'a','a');
```

Before the PR, they may get RowID as:
```
thread1:  1, 3;
thread2: 2, 4;
```

After the PR, they will get RowID as:
```
thread1:  1, 2;
thread2: 3, 4;
```

### What is changed and how it works?

I will allocate a continuous numbers every time a statement try to get row id.

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change

Side effects

 - Increased code complexity

Related changes

No

Release note

 - The record in the same statement would get continuous RowID. You could use `SELECT _tidb_rowid for TABLE` to ensure this feature.

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
Please describe the benefits for continuous row id.

File: table/tables/tables.go [hunk 1]
```diff
@@ -455,9 +455,22 @@ func (t *tableCommon) AddRecord(ctx sessionctx.Context, r []types.Datum, opts ..
...
+		stmtCtx := ctx.GetSessionVars().StmtCtx
+		rows := stmtCtx.RecordRows()
+		if rows > 1 {
+			if stmtCtx.BaseRowID >= stmtCtx.MaxRowID {
+				stmtCtx.BaseRowID, stmtCtx.MaxRowID, err = t.AllocHandleIDs(ctx, rows)
+				if err != nil {
+					return 0, err
+				}
+			}
+			stmtCtx.BaseRowID += 1
```
Comment (zhangjinpeng1987) [inline table/tables/tables.go]: Should increase BaseRowID after L468?

File: table/tables/tables.go [hunk 2]
```diff
@@ -947,9 +960,15 @@ func GetColDefaultValue(ctx sessionctx.Context, col *table.Column, defaultVals [
...
 // AllocHandle implements table.Table AllocHandle interface.
 func (t *tableCommon) AllocHandle(ctx sessionctx.Context) (int64, error) {
-	_, rowID, err := t.Allocator(ctx).Alloc(t.tableID, 1)
+	_, rowID, err := t.AllocHandleIDs(ctx, 1)
+	return rowID, err
+}
+
+// AllocHandle implements table.Table AllocHandle interface.
+func (t *tableCommon) AllocHandleIDs(ctx sessionctx.Context, n uint64) (int64, int64, error) {
+	base, rowID, err := t.Allocator(ctx).Alloc(t.tableID, n)
```
Comment (zhangjinpeng1987) [inline table/tables/tables.go]: s/rowID/maxID

**Comment by Little-Wallace** [comment]
image

This PR could reduce more than 15% CPU of TiKV server when insert many keys.

The query `insert 1` is  `sysbench prepare`.
The query  `insert 2` is `batch_insert` of benchbot. See it at (https://github.com/pingcap/productivity-tools/blob/master/benchmark/batch_insert/batch_insert.lua#L75)

File: table/tables/tables.go [hunk 3]
```diff
@@ -455,9 +455,22 @@ func (t *tableCommon) AddRecord(ctx sessionctx.Context, r []types.Datum, opts ..
...
 	}
 	if !hasRecordID {
-		recordID, err = t.AllocHandle(ctx)
-		if err != nil {
-			return 0, err
+		stmtCtx := ctx.GetSessionVars().StmtCtx
+		rows := stmtCtx.RecordRows()
+		if rows > 1 {
+			if stmtCtx.BaseRowID >= stmtCtx.MaxRowID {
+				stmtCtx.BaseRowID, stmtCtx.MaxRowID, err = t.AllocHandleIDs(ctx, rows)
```
Comment (winoros) [inline table/tables/tables.go]: The `MaxRowID` is not used actually?

File: executor/insert_test.go [hunk 4]
```diff
@@ -763,6 +765,43 @@ func (s *testSuite3) TestBit(c *C) {
...
+			tk := testkit.NewTestKitWithInit(c, s.store)
+			for j := 0; j < 10; j++ {
+				k := strconv.Itoa(idx*100 + j)
+				sql := "insert into t1(a,b) values (" + k + ", 2)"
+				for t := 0; t < 20; t++ {
+					sql += ",(" + k + ",2)"
+				}
+				tk.MustExec(sql)
+				q := "select _tidb_rowid from t1 where a=" + k
+				fmt.Printf("query: %v\n", q)
```
Comment (tiancaiamao) [inline executor/insert_test.go]: .......

Comment (zhangjinpeng1987) [reply executor/insert_test.go]: @Little-Wallace

File: table/tables/tables.go [hunk 5]
```diff
@@ -455,9 +455,22 @@ func (t *tableCommon) AddRecord(ctx sessionctx.Context, r []types.Datum, opts ..
 		}
 	}
 	if !hasRecordID {
-		recordID, err = t.AllocHandle(ctx)
-		if err != nil {
-			return 0, err
+		stmtCtx := ctx.GetSessionVars().StmtCtx
+		rows := stmtCtx.RecordRows()
```
Comment (tiancaiamao) [inline table/tables/tables.go]: This number is wrong...
It's not the row counts in this statement, it's a batch of chunk rows and accumulative.

**Pull request merged**